### ISSUE TITLE
Reflection of computed column

### ISSUE THREAD

**Issue by CaselIT** [pauldex/sqlalchemy-firebird#4]
I've seen that this is replacing the internal firebird dialect in sqlalchemy.

I'm adding support for reflecting computed columns in https://github.com/sqlalchemy/sqlalchemy/pull/5064.
Should the support be added in that pr also for firebird, or should a new pr be made here?

**Comment by pauldex** [comment]
I think we should minimize (eliminate?) custom code to support specific dialects in SQLAlchemy.  It would be more flexible and easier to maintain to have the plug-in dialect provide the needed support, especially since the long-range goal to have every database provider as a plug-in.

 

What do you think?

Paul

**Comment by CaselIT** [comment]
Yes I agree, that's the reason why I've opened this issue.
Once the referenced pr is merged support for it could be added here.

**Comment by gordthompson** [comment]
@CaselIT 

> Once the referenced pr is merged support for it could be added here.

Now that it's merged I thought I'd see what happened when I enabled `computed_columns`

```diff
Index: sqlalchemy_firebird/requirements.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sqlalchemy_firebird/requirements.py	(revision 807c47c244d3a531abd6430808914e825ed0b5db)
+++ sqlalchemy_firebird/requirements.py	(date 1586010192231)
@@ -8,6 +8,11 @@
     def autoincrement_insert(self):
         return exclusions.closed()
 
+    @property
+    def computed_columns(self):
+        "Supports computed columns"
+        return exclusions.open()
+
     @property
     def datetime_microseconds(self):
         return exclusions.closed()
```

and then ran

```
pytest -k ComputedReflectionTest --db firebird2 -s --log-debug=sqlalchemy.engine | tee computed_reflect.txt
```

but the tests all error out because they seem to be trying to create the test tables after they already exist.

[computed_reflect.txt](https://github.com/pauldex/sqlalchemy-firebird/files/4431850/computed_reflect.txt)

**Comment by CaselIT** [comment]
The reflection test does seem to create two table with the same name in different schemas:
https://github.com/sqlalchemy/sqlalchemy/blob/c7d3ca0da477451885158a923aa9ee7e49794541/lib/sqlalchemy/testing/fixtures.py#L458-L473

does firebird not support schemas?

**Comment by gordthompson** [comment]
> does firebird not support schemas?

It appears not.

```python
class TableDDLTest(_TableDDLTest):
    def test_create_table_schema(self):
        """Do not test schemas

        In Firebird, a schema is the same thing as a database.  According to the Firebird
        reference manual, "The CREATE DATABASE statement creates a new database. You can use
        CREATE DATABASE or CREATE SCHEMA. They are synonymous."  See:
        https://firebirdsql.org/file/documentation/reference_manuals/fblangref25-en/html/fblangref25-ddl-db.html
        """
        return
```

**Comment by CaselIT** [comment]
Mysql seems to be the same, but the test work on it.
Maybe these is some other setting to change to enable schema to run as in mysql?

**Comment by gordthompson** [comment]
`CREATE DATABASE/SCHEMA` is something we haven't yet addressed for this dialect. I did some preliminary investigation a while ago and came to the conclusion that it might be messy. Trying to `CREATE SCHEMA` via `fdb` (DBAPI layer) threw a "Cannot prepare ..." error. Creating from the command-line `isql` tool worked on Windows but required `sudo` on Linux because of a central security-type database that lives in a folder owned by `root`. My conclusion was that it *might* be possible with sufficient fiddling of permissions, but that was a rabbit hole I chose to avoid, at least for the time being.

Interestingly enough, SQLAlchemy does have a `schemas` requirement that appears to be closed by default

https://github.com/sqlalchemy/sqlalchemy/blob/c7d3ca0da477451885158a923aa9ee7e49794541/lib/sqlalchemy/testing/requirements.py#L391-L396

and we haven't opened it here.

**Comment by CaselIT** [comment]
I guess the tests need to be updated on the sqlalchemy side to take into account the schema requirement then.

**Comment by gordthompson** [comment]
As it turns out, the tests were just being skipped for MySQL.

Now when I run the tests for this dialect, using the patched version of SQLAlchemy from ...

https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1841

... the setup errors are gone and one of the tests passes ...

```
============================= test session starts =============================
platform win32 -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- c:\users\gord\git\sqlalchemy-firebird\venv\scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\Gord\git\sqlalchemy-firebird, inifile: setup.cfg
plugins: forked-1.1.3, xdist-1.31.0
collecting ... collected 379 items / 375 deselected / 4 selected

test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_computed_col_default_not_set PASSED [ 25%]
test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_computed FAILED [ 50%]
test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_persisted FAILED [ 75%]
test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_persisted_with_schama FAILED [100%]

================================== FAILURES ===================================
_ ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird.test_get_column_returns_computed _
Traceback (most recent call last):
  File "c:\users\gord\git\sqla-gerrit\lib\sqlalchemy\testing\suite\test_reflection.py", line 1204, in test_get_column_returns_computed
    is_true("computed" in compData)
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 219, in is_true
    is_(a, True, msg=msg)
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 228, in is_
    assert a is b, msg or "%r is not %r" % (a, b)
AssertionError: False is not True
assert False is True
_ ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird.test_get_column_returns_persisted _
Traceback (most recent call last):
  File "c:\users\gord\git\sqla-gerrit\lib\sqlalchemy\testing\suite\test_reflection.py", line 1235, in test_get_column_returns_persisted
    testing.requires.computed_columns_default_persisted.enabled,
  File "c:\users\gord\git\sqla-gerrit\lib\sqlalchemy\testing\suite\test_reflection.py", line 1218, in check_column
    is_true("computed" in data[column])
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 219, in is_true
    is_(a, True, msg=msg)
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 228, in is_
    assert a is b, msg or "%r is not %r" % (a, b)
AssertionError: False is not True
assert False is True
_ ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird.test_get_column_returns_persisted_with_schama _
Traceback (most recent call last):
  File "c:\users\gord\git\sqla-gerrit\lib\sqlalchemy\testing\suite\test_reflection.py", line 1258, in test_get_column_returns_persisted_with_schama
    testing.requires.computed_columns_default_persisted.enabled,
  File "c:\users\gord\git\sqla-gerrit\lib\sqlalchemy\testing\suite\test_reflection.py", line 1218, in check_column
    is_true("computed" in data[column])
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 219, in is_true
    is_(a, True, msg=msg)
  File "C:\Users\Gord\git\sqla-gerrit\test\..\lib\sqlalchemy\testing\assertions.py", line 228, in is_
    assert a is b, msg or "%r is not %r" % (a, b)
AssertionError: False is not True
assert False is True
=========================== short test summary info ===========================
FAILED test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_computed
FAILED test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_persisted
FAILED test/test_suite.py::ComputedReflectionTest_firebird2+fdb_3_0_33054_firebird::test_get_column_returns_persisted_with_schama
================= 3 failed, 1 passed, 375 deselected in 2.67s =================
```

... so that's progress, at least.

**Comment by CaselIT** [comment]
Nice. There are 4 requirement flag in the requirements related to the computed columns.
```
computed_columns_stored
computed_columns_virtual
computed_columns_default_persisted
computed_columns_reflect_persisted
```
maybe some of them are set wrong. Firebird is set in the requirements of sqlalchemy
